### PR TITLE
Vendor implementation

### PR DESCRIPTION
**Pull request by Matt-Hurd**
* Selling to vendor
* Buying from vendor
* Buyback

Important TODOs:
* Figure out item cost calculation from GameFormula

### CODE REVIEW

File: Source/NexusForever.WorldServer/Network/Message/Model/ClientVendorSell.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    [Message(GameMessageOpcode.ClientSellItemToVendor, MessageDirection.Client)]
+    public class ClientVendorSell : IReadable
+    {
+        public ItemLocation ItemLocation { get; } = new ItemLocation();
+        public uint Quantity { get; private set; }
+
+        public void Read(GamePacketReader reader)
+        {
+            ItemLocation.Read(reader);
+            Quantity = (uint)reader.ReadInt();
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Model/ClientVendorSell.cs]: There a reason you can't just call ReadUInt here?

File: Source/NexusForever.WorldServer/Network/Message/Model/ServerBuybackItemsUpdated.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        public BuybackItem BuybackItem { get; set;}
+        public ulong Unk3 { get; set; } = 0;
+        public uint Unk4 { get; set; } = 0;
+        public ulong Unk5 { get; set; } = 0;
+        public ulong Unk6 { get; set; } = 0;
+        public uint Unk7 { get; set; } = 0;
+        public byte[] Unk8 { get; set; } = new byte[20];
+        public byte[] Unk9 { get; set; } = new byte[32];
+        public uint UnkE { get; set; } = 0;
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Model/ServerBuybackItemsUpdated.cs]: There is no need to specify integer defaults of 0

File: Source/NexusForever.WorldServer/Network/Message/Handler/VendorHandler.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                return;
+
+            if (currency0 != null)
+                session.Player.CurrencyManager.CurrencySubtractAmount(currency0.Entry, calculatedCost0);
+
+            if (currency1 != null)
+                session.Player.CurrencyManager.CurrencySubtractAmount(currency1.Entry, calculatedCost1);
+
+            var item = new Game.Entity.Item(session.Player.CharacterId, itemEntry, Math.Min(1, itemEntry.MaxStackCount));
+            session.Player.Inventory.ItemCreate(itemEntry.Id, vendorPurchase.VendorItemQty * itemEntry.BuyFromVendorStackCount);
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Handler/VendorHandler.cs:97]: Be aware that this won't work in all cases.
Items that has specific attributes assigned to them will be lost on repurchase.

File: Source/NexusForever.WorldServer/Network/Message/Handler/VendorHandler.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            EntityVendorItem vendorItem = VendorInfo.GetItemAtIndex(vendorPurchase.VendorIndex);
+            if (vendorItem == null)
+                return;
+
+            Item2Entry itemEntry = GameTableManager.Item.GetEntry(vendorItem.ItemId);
+            float costMultiplier = vendorPurchase.VendorItemQty * VendorInfo.BuyPriceMultiplier;
+            Currency currency0 = session.Player.CurrencyManager.GetCurrency(itemEntry.CurrencyTypeId0);
+            Currency currency1 = session.Player.CurrencyManager.GetCurrency(itemEntry.CurrencyTypeId1);
+            ulong calculatedCost0 = (ulong)(itemEntry.CurrencyAmount0 * costMultiplier);
+            ulong calculatedCost1 = (ulong)(itemEntry.CurrencyAmount1 * costMultiplier);
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Handler/VendorHandler.cs]: I would suggest using the GameTableFieldArrayAttribute here to remove the need to check both currencies explicitly

File: Source/NexusForever.WorldServer/Game/Entity/Player.cs [hunk 5]
```diff
@@ -46,6 +46,9 @@ public byte Level
         public Inventory Inventory { get; }
         public CurrencyManager CurrencyManager { get; }
         public WorldSession Session { get; }
+        public VendorInfo SelectedVendorInfo { get; set; } // TODO unset this when too far away from vendor
+        public Dictionary<uint, BuybackItem> BuybackItems = new Dictionary<uint, BuybackItem>();
+        public uint highestBuybackId = 0; // TODO Clean this implementation
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/Player.cs]: Maybe just iterate BuybackItems and find the highest BuybackItemId and +1?

**Review by Rawaho — CHANGES_REQUESTED** [review]
Just a few issues that need addressing.

File: Source/NexusForever.WorldServer/Game/Entity/Player.cs (hunk 5 above)
Comment (Matt-Hurd) [reply Source/NexusForever.WorldServer/Game/Entity/Player.cs]: The game expects a unique buybackId to be sent. ie:
BuybackItems 1-5 are available, you buy 5, the next one should still be 6.

**Review by Rawaho — APPROVED** [review]

**Pull request merged**